### PR TITLE
Issue 9 db read all tables

### PR DESCRIPTION
**Pull request by 0xd5dc**
I fixed the Db for issue 9.
Db's load() reads all tables accordingly from a sqlite file if pre_existing_keys or table_names given.

ps, Db can do Db_sqlalchemy's job as well

### CODE REVIEW

**Comment by milicag** [comment]
Thank you for starting to address this issue. This is what I get on this branch, followed by a pass on master:
```
ERROR: test_load_pre_keys (adapter.tests.test_to_python.TestDb)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/mgrahovac/repos/public/Adapter/adapter/tests/test_to_python.py", line 99, in test_load_pre_keys
    self.assertEqual(len(self.db.load()), 2)
  File "/Users/mgrahovac/repos/public/Adapter/adapter/to_python.py", line 247, in load
    raise ImportError(f'Cannot find {self.file_path}')
ImportError: Cannot find ./test.db

----------------------------------------------------------------------
Ran 7 tests in 0.602s

FAILED (errors=4)
(atcd) mgrahovac@milicag Adapter % git checkout master
Switched to branch 'master'
Your branch is up to date with 'origin/master'.
(atcd) mgrahovac@milicag Adapter % python -m unittest adapter.tests.test_to_python
..INFO:adapter.to_python:Connected to: adapter/tests/test.xlsx
Read in <Table 'xlsx_table1': test!$J$2:$L$5>
Read in <Table 'xlsx_table2': test!$J$9:$L$12>
Read in <Table 'run_parameters': test!$B$2:$C$3>
Read in <Table 'xlsx_single_col_table': test!$L$21:$L$25>
Read in <Name 'xlsx_named_range1': =test!$J$16:$K$18>
Read in <Name 'xlsx_single_col_range': =test!$J$21:$J$25>
Read in <Name 'xlsx_single_value_named_range': =test!$J$29>
INFO:adapter.to_python:Read in input tables and ranges from adapter/tests/test.xlsx.
Read in <Table 'xlsx_table1': test!$J$2:$L$5>
Read in <Table 'xlsx_table2': test!$J$9:$L$12>
INFO:adapter.to_python:Read in input tables and ranges from adapter/tests/test.xlsx.
Read in <Name 'xlsx_named_range1': =test!$J$16:$K$18>
Read in <Name 'xlsx_single_col_range': =test!$J$21:$J$25>
INFO:adapter.to_python:Read in input tables and ranges from adapter/tests/test.xlsx.
.
----------------------------------------------------------------------
Ran 3 tests in 0.656s

OK
```

**Comment by milicag** [comment]
@0xd5dc Also please make sure to test it with ATCD base tools on our end.

**Comment by evaneill** [comment]
I'm also getting errors about the missing test .db files when running the `test_to_python`. If they're really small I'd just commit them. Otherwise instantiating them could be added to the setup.

But this is working for me running NIA & shipments unit tests, so I think it's working! I also ran some timing tests on a small ~30MB local database, querying a ~1MB table, and this was 30x faster than master (makes sense given the size). Exciting improvement!

A few smaller things:
1. you can increment the patch version in adapter/__init__ before you eventually merge & tag
2. The IO error about (# tables in db == 0) can include the self.file_path for easier debugging
3. When a value in self.pre_existing_keys is in table_names, I think it should still throw an error (the implication being that adapter has already loaded a table with this name before, and there's no good way to know which it should prefer). Can use Debugger to check & throw the error

Thanks for doing this!

**Comment by 0xd5dc** [comment]
@milicag I think you tested it using the master branch. I did test it with one of the ATCD base tools successfully.
```
pip uninstall adapterio 
pip install git+https://github.com/LBNL-ETA/Adapter.git@issue_9_db_read_all_tables
```
@evaneill You're welcome! I will update the version, will include the file path in the error message. 
What's the purpose of Debugger?

**Comment by milicag** [comment]
@0xd5dc no, I tested your branch and the master as provided in my comment that I 🤔 you are referring to.

**Comment by 0xd5dc** [comment]
I think it's ready for merge but last test with one of the ATCD base tools failed because of value error while handing exception.
@evaneill Can you test with your tools to see if the new commit works?

**Comment by evaneill** [comment]
The tests still work for me with our tools with this adapter. 

I also got these tests to pass, but ONLY if I run them from the test directory itself. Ideally want to be able to run them from root. Can probably just define a constant and use it in the code. 

`TEST_DIR = os.path.dirname(__file__)` and then wherever the test files are read, do an `os.path.join(TEST_DIR,...)`

doing the os.path.join will also help since conventions like './test.db' won't work on windows

**Comment by evaneill** [comment]
The tests still work for me with our tools with this adapter. 

I also got these tests to pass, but ONLY if I run them from the test directory itself. Ideally want to be able to run them from root. Can probably just define a constant at the top of test files (e.g. `test_to_python.py`) and use it in the code:

`TEST_DIR = os.path.dirname(__file__)` and then wherever the test files are read, do an `os.path.join(TEST_DIR,...)`

doing the os.path.join will also help since conventions like './test.db' won't work on windows

File: adapter/to_python.py [hunk 1]
```diff
@@ -222,68 +216,60 @@ def get_named_data_objects(self, data_object_names, kind="all"):
...
-        """Loads tables
+        """Loads tables from a sqlite file
 
         Parameters:
 
             table_names: list
-                List of table names to load.
+                List of table names to load. Assuming the given tables always exist in the db file
                 Default: None = load all tables
+        Notes
```
Comment (milicag) [inline adapter/to_python.py:244]: @0xd5dc is this now resolved or still pending? I don't understand the issue.

**Review by milicag — COMMENTED** [review]
Hi @0xd5dc - please take a look at my comment. Excuse my omission in the previous review pass. Let me know if there are questions.

File: adapter/to_python.py (hunk 1 above)
Comment (0xd5dc) [reply adapter/to_python.py:244]: the section was comment out because other apps don't have any handler for IO error.

Comment (milicag) [reply adapter/to_python.py:244]: The adapter uses the logger that is used throughout our tools, please look at any of the log files for a test run. With that, unless I'm missing something, this note can be removed. If you need help with understanding how our logger works, please check in with me or someone else from the team.

File: adapter/to_python.py [hunk 2]
```diff
@@ -222,68 +216,56 @@ def get_named_data_objects(self, data_object_names, kind="all"):
...
+            raise ImportError(f'Cannot find {self.file_path}')
+        con_str = f'sqlite+pysqlite:///{self.file_path}'
+        engine = create_engine(con_str)
+        metadata = MetaData()
+        metadata.reflect(bind=engine)
+
+        keys = metadata.tables.keys()
+        if len(keys) == 0:
+            # check database integrity
+            raise IOError('0 table found in the database file! The input file may be unsupported or corrupted')
```
[the remark this answers is not in the document]
Comment (0xd5dc) [reply adapter/to_python.py]: Noted. By the way, this code was outdated.

[the remark this answers is not in the document]
Comment (milicag) [reply adapter/to_python.py]: What does that mean? Is it on the current master?

[the remark this answers is not in the document]
Comment (milicag) [reply adapter/to_python.py]: I see, it is not. Ok, thanks.

File: adapter/to_python.py (hunk 1 above)
Comment (milicag) [reply adapter/to_python.py:244]: Of course, if you have ideas on improving the logging, please feel free to implement them. I would say sticking with the logger as our approach to logging is useful as we have it throughout and it allows us to create clear log files, especially the approach Evan took recently.

**Pull request merged**